Modify the updating process & Add documentation

v-bertin opened the pull request:
Modify the updating process
======================
The first commit solves the following issue : "The container systemd service files is not recopied from the /apps/container... directory into the /etc/systemd/system. It would be nice to do it to be able to update that file too."
The previous updating process was made in this order : copy the service file, start the associated container, notify the Hawkbit server and do so until all containers have been updated. 
This current implementation leads to 2 main issue:

1. Dependency issues among systemd service files 
2. Server notified with a "update successful" notification after first container being updated. Its state would not change afterwards even if the update of the following container failed. This issue appears as soon as we try to update more than one app / container.

The new implementation allows to get rid of these problems with a new order : copy all service files, regenerate systemd dependency tree with a method equivalent to "systemctl daemon-reload", start all containers, and according to the result of the update process, notify the server with a "update successful" or a "update failure" notification.

Add documentation
================
The second commit adapts the documentation syntax to fit the reST standards of a Sphinx Project.
The third commit adds Sphinx project specific files and tips on how to generate locally the FullMetalUpdate documentation.
The fourth commit fixes some issues in Sphinx documentation generation.

Code Cleaning
===============
The fifth, sixth and seventh commits clean the code and allow to fix some problems in the new implementation of the updating process.

Fix notification system within the scope of the updating process
===================================================
This part is related to the eighth, the ninth and the tenth commits. As the functioning of containers implementing the notify feature of systemd is a bit different, I had to adapt the new process.
From now on, feedback threads don't feedback the Hawkbit server themselves but they pass the information (the startup was a success, the startup was a failure) to the main thread (FMU client). Plus, each feedback thread is associated with its own socket, allowing each notify container to be monitored by its own thread.
When all containers have been restarted, the client finally makes the feedback to the server and precises which containers couldn't start or couldn't be updated.

v-bertin committed:
Modify the updating process to solve issues bound to dependancies and to updates of multiple containers

v-bertin committed:
Review, improve and change syntax of Python documentation (reST syntax for Sphinx projects)

v-bertin committed:
Add Sphinx Project and modify README with explanations on how to build the documentation

v-bertin committed:
Last update of the Sphinx Documentation

v-bertin committed:
Add restart of containers after rollback to deal with the fact that new update_container() doesn't start containers itself

v-bertin committed:
Remove unused method install() and unused attribute lock_keeper

v-bertin committed:
Fix invalid calls of update_container() method

v-bertin committed:
Remove disruptive feedbacks generated by feedback threads

The Hawkbit server must only be feedbacked once at the end of the whole updating process.

v-bertin committed:
Correct the feedback process for notify containers

We now use 1 Unix Socket per feedback thread, knowing that each feedback thread is associated with a unique notify container.
It allows a better monitoring of transactions bewteen notify containers (senders) and feedback threads (receivers).

Moreover, we now start all containers at the same time : so everything (= start + acknoledgement of the success or failure of container startup) fits in the 35s during which the connexion is opened between the feedback thread and the socket.

v-bertin committed:
Add safety mutex to protect results dictionnary from concurrent accesses

v-bertin committed:
Fix a few mistakes

jvial-witekio commented on fullmetalupdate/fullmetalupdate_ddi_client.py:
```diff
@@ -93,14 +114,28 @@ async def cancel(self, base):
+            - Notifies Hawkbit server about the appropriate start of the update ;
+            - All chunks are then parsed, ie 
+                1) OS chunk are processed and cause a system reboot ;
+                2) Container Chunk cuase container updates ;
```
Typo "cuase"

jvial-witekio commented on fullmetalupdate/fullmetalupdate_ddi_client.py:
```diff
@@ -93,14 +114,28 @@ async def cancel(self, base):
+            - All chunks are then parsed, ie 
+                1) OS chunk are processed and cause a system reboot ;
+                2) Container Chunk cuase container updates ;
+            - Systemd dependancy tree is regenerated in order to take into account potential changes in dependancies caused by new service files ;
```
dependancy instead of dependency
dependancies instead of dependencies

In fact the goal is to take into account every change in the service files

jvial-witekio commented on doc/index.rst:
```diff
@@ -37,11 +36,18 @@ Technical documentation
+   ./src/bootup.md
    ./src/add-targets-to-hawkbit.md
 
+| See the `Rollback Feature <https://github.com/v-bertin/documentation/blob/master/rollback.md>`_ documentation on Github.
```
Point on a file of your Github repo

jvial-witekio commented on fullmetalupdate/fullmetalupdate_ddi_client.py:
```diff
@@ -447,55 +451,44 @@ def container_feedbacker(self,
-                    msg = "The notify container started successfully"
-                    status_result = DeploymentStatusResult.success
-                    status_execution = DeploymentStatusExecution.closed
+                    msg = "Container" + container_name + " started successfully"
```
Missing space after at the end of "Container" ?

v-bertin committed:
Code review #1

Fix typo issues in Python documentation.
Fix invalid link to personal Github repo.
